[C10D] Fix coalescedCollective op Flight Recording

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120430


Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang

wconstab committed:
working on coalesced

[ghstack-poisoned]

wconstab committed:
Update on "working on coalesced"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab committed:
Update on "Make coalescedCollective work with FlightRecorder"




[ghstack-poisoned]

wconstab committed:
Update on "Make coalescedCollective work with FlightRecorder"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2607,11 +2605,10 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collectiveCoalesced(
-  } else {
-    at::cuda::CUDAGraph::dec_pending_event_queries();
-  }
+  workEnqueue(work);
+  // TODO(whc) did I delete a relevant codepath here?
+  // at::cuda::CUDAGraph::dec_pending_event_queries();
+  // }
```
Yes, I think you would need to put back the `capture_status == c10::cuda::CaptureStatus::None` check and the `else` path.

Would be also nice to pack everything:
```
inc_pending_event_queries()
check capture_status
dec_pending_event_queries()
```
inside `workEnqueue` so that we don't have these checks everywhere. (can be in a different PR)

wconstab replied:
I think my reasoning here was that coalescing_state_ should NEVER be active inside this function.  Becuase this function is part of the 'coalescing manager' implementation, which would call its own startCoalescing function later.

If that is correct, i'll add an assertion here to clarify it.

wconstab replied:
i'm not sure how the cudagraph capture path works.  Is it exclusive with coalescing or is it additive?

should the new logic be like this?  I don't understand why the cuda event increment is unconditional and the decrement is conditional.

```
  // Notify graphs before we check the capture status preemptively
  at::cuda::CUDAGraph::inc_pending_event_queries();

  workEnqueue(work);
  if (capture_status != c10::cuda::CaptureStatus::None) {
    at::cuda::CUDAGraph::dec_pending_event_queries();
  }
```

kwen2501 replied:
Sorry for the confusion. 

What I meant is like this:
```
  at::cuda::CUDAGraph::inc_pending_event_queries();

  if (capture_status == c10::cuda::CaptureStatus::None) {
    workEnqueue(work);
  } else {
    at::cuda::CUDAGraph::dec_pending_event_queries();
  }
```

It preemptively increasing the pending work number, if this work is actually enqueued, then this is fine.
If this work ends up not being enqueued because CUDA Graph capture is active, then it winds back the pending work number.

kwen2501 replied:
You can check elsewhere in the code how `workEnqueue(work)` is conditioned. I guess they are more or less done the same way. Hence I propose putting the above CUDA Graph logic into the `workEnqueue` function itself, so that we don't have the if-else at every call site.

wconstab replied:
hmm.

is it necessary to preemptively increment and then decrement?  or can we just increment when we workEnqueue nad never decrement?

kwen2501 replied:
There was a reason. See here:
https://github.com/pytorch/pytorch/pull/110665#issuecomment-1751403208
Cc @eqy

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

kwen2501 reviewed (approved):
Thanks for the last fix about workEnqueue. LGTM.

eqy commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2622,14 +2622,43 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collectiveCoalesced(
+  we also change the semantic of workEnqueue to 'maybeWorkEnqueue'.
+
+  TODO:
+   - Are these assumptions correct? (just pieced this together following
```
Yes, I think this explanation is accurate; one nit would maybe be editing "(a) capturing is already in progress, so we can not enqueue" to "(a) capturing is already in progress (we cannot enqueue in this case)."

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "[C10D] Fix coalescedCollective op Flight Recording"



Also noticed and filed https://github.com/pytorch/pytorch/issues/120516 during this work. May land this as is and then test/fix the other varieties of coalesced collectives later.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l yf225

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
